### ISSUE TITLE
systemd-fstab-generator skips some mountpoints and mounts root as read-only

### ISSUE THREAD

**Issue by aik** [systemd/systemd#26847]
### systemd version the issue has been seen with

249.11-0ubuntu3.7

### Used distribution

Armbian 22.11.4 Jammy

### Linux kernel version used

Linux 5.15.93-sunxi64

### CPU architectures issue was seen on

aarch64

### Expected behaviour you didn't see

systemd-fstab-generator produces incorrect results. zramfs disables, tmpfiles disabled.

```console
$ cat /etc/fstab 
UUID=ecc7fb6d-12ab-41b9-b11a-c2c9b78e3bd1 / ext4 defaults,noatime 0 1
/dev/sda2 none swap defaults 0 0
/dev/sda3 /mnt/sda ext4 defaults,sync 0 0
/dev/sda4 /tmp ext4 defaults,sync 0 0
```

So I expect all the above mounted.

### Unexpected behaviour you saw

The root is mounted read-only and /tmp is not mounted at all. /tmp has this:

```console
$ ls
armbian-hardware-optimization.Y1wcWo  local-fs.target.requires  mnt-sda.mount  mystrace              tmp.mount
dev-sda2.swap                         local-fs.target.wants     -.mount        swap.target.requires  tmux-1000
$ cat dev-sda2.swap 
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
After=blockdev@dev-sda2.target

[Swap]
What=/dev/sda2
aik@hass:/tmp$ cat mnt-sda.mount 
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
Before=local-fs.target
After=blockdev@dev-sda3.target

[Mount]
Where=/mnt/sda
What=/dev/sda3
Type=ext4
Options=defaults,sync
aik@hass:/tmp$ cat ./-.mount
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
Before=local-fs.target
aik@hass:/tmp$ cat tmp.mount 
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
Before=local-fs.target
After=blockdev@dev-sda4.target

[Mount]
Where=/tmp
What=/dev/sda4
Type=ext4
Options=defaults,sync

```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/system-generators/systemd-fstab-generator produces

root@hass:/tmp# tail -n 9 /var/log/syslog 
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Found container virtualization none.
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Parsing /etc/fstab...
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Found entry what=/dev/disk/by-uuid/ecc7fb6d-12ab-41b9-b11a-c2c9b78e3bd1 where=/ type=ext4 makefs=no growfs=no noauto=no nofail=no
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Failed to create symlink /tmp/local-fs.target.wants/systemd-fsck-root.service: File exists
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Found entry what=/dev/sda2 where=none type=swap makefs=no growfs=no noauto=no nofail=no
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: SELinux enabled state cached to: disabled
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Found entry what=/dev/sda3 where=/mnt/sda type=ext4 makefs=no growfs=no noauto=no nofail=no
Mar 16 11:11:58 hass systemd-fstab-generator[4734]: Found entry what=/dev/sda4 where=/tmp type=ext4 makefs=no growfs=no noauto=no nofail=no
```

This is recorded right after booting:
```console
$ cat /etc/fstab 
UUID=ecc7fb6d-12ab-41b9-b11a-c2c9b78e3bd1 / ext4 defaults,noatime 0 1
/dev/sda2 none swap defaults 0 0
/dev/sda3 /mnt/sda ext4 defaults,sync 0 0
/dev/sda4 /tmp ext4 defaults,sync 0 0
$ cat /proc/mounts 
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=431228k,nr_inodes=107807,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,nodev,noexec,relatime,size=100872k,mode=755 0 0
/dev/mmcblk0p1 / ext4 ro,relatime 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
bpf /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,nosuid,nodev,noexec,relatime 0 0
sunrpc /run/rpc_pipefs rpc_pipefs rw,relatime 0 0
nfsd /proc/fs/nfsd nfsd rw,relatime 0 0
/dev/mmcblk0p1 /var/log.hdd ext4 ro,relatime 0 0
armbian-ramlog /var/log tmpfs rw,nosuid,nodev,noexec,relatime,size=51200k,mode=755 0 0
fusectl /sys/fs/fuse/connections fusectl rw,nosuid,nodev,noexec,relatime 0 0
configfs /sys/kernel/config configfs rw,nosuid,nodev,noexec,relatime 0 0
/dev/sda3 /mnt/sda ext4 rw,sync,relatime 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=100868k,nr_inodes=25217,mode=700,uid=1000,gid=1000 0 0
$ df
Filesystem     1K-blocks     Used Available Use% Mounted on
tmpfs             100872     3212     97660   4% /run
/dev/mmcblk0p1  30254472  8308276  21605500  28% /
tmpfs             504352        0    504352   0% /dev/shm
tmpfs               5120        4      5116   1% /run/lock
armbian-ramlog     51200     2060     49140   5% /var/log
/dev/sda3      102626232 10219024  87147944  11% /mnt/sda
tmpfs             100868        0    100868   0% /run/user/1000
$ ls /tmp/
$ sudo mount -o remount /
$ sudo mount /tmp
$ df
Filesystem     1K-blocks     Used Available Use% Mounted on
tmpfs             100872     3212     97660   4% /run
/dev/mmcblk0p1  30254472  8308276  21605500  28% /
tmpfs             504352        0    504352   0% /dev/shm
tmpfs               5120        4      5116   1% /run/lock
armbian-ramlog     51200     2060     49140   5% /var/log
/dev/sda3      102626232 10219176  87147792  11% /mnt/sda
tmpfs             100868        0    100868   0% /run/user/1000
/dev/sda4       15559884      136  14747548   1% /tmp
```

**Comment by keszybz** [comment]
As noted in the issue template, we only support the latest two releases in the upstream bugtracker, and we're currently at v253.  Please discuss this issue with your distro maintainers.

It seems to be an operator error: `systemd-fstab-generator` (and all our generators) would write to `/tmp/` if the output directory was not specified. When pid1 calls the generators, it always does that with a set of output directories. You can call it by hand, for debugging, but you have to make sure that the output directory is fresh and empty before each run. (Hence the default of `/tmp` is not terribly useful, and has already been removed upstream.)  The messages you are seeing are caused by calling the generator twice on the same (implicit) output directory, and are not related to how the generator behaves when it is invoked by pid1.

tl;dr: we don't know what happened during boot and why.

**Comment by keszybz** [comment]
The log parts of which you pasted contains the information about what happened and when. And the dependencies are listed in the unit files and can also be queried at runtime over dbus or using command line tools.
